### DISCUSSION TITLE
Malformed Open Graph/Twitter Cards Image meta

### DISCUSSION THREAD

**Discussion by TetraTheta** [hbstack/.github#33] [Q&A] [answered]
Sorry, I couldn't find out where to put this issue at.

My expectation for Open Graph and Twitter Cards are like these:
* If `images` is defined in post front matter, it should be used as `og:image` and `twitter:image`.
  * Many post would have `images` defined in its front matter, so it will be used.
* If the page doesn't have any front matter or `images` is not defined in front matter, `params.images` will be used instead.
  * This includes taxonomy pages and blog main page.

I always write post front matter like this:
```yaml
---
images:
  - 001.webp
---
```
It should refer to `001.webp` which is at same location where `index.md` is at. At least, it worked nicely for creating 'blog post card' in my blog.

But today, I found that I don't see any 'preview image' in Discord's link preview embed. Discord uses Open Graph (or Twitter Cards) to create fields of embed, so I inspected my blog's meta with [opengraph.xyz](https://www.opengraph.xyz).

According to Official Hugo Documentation, Hugo supports both [Open Graph](https://gohugo.io/templates/internal/#open-graph) and [Twitter Cards](https://gohugo.io/templates/internal/#twitter-cards).

## Post pages
When I inspected meta with [one of my blog post link](https://tetralog.haipa.xyz/blog/genshin-impact/misc/re-adventure-rank-30/), I found `og:image` and `twitter:image` are pointing to wrong path.
```html
<meta property="og:image" content="http://tetralog.haipa.xyz/001.webp">
<meta name="twitter:image" content="http://tetralog.haipa.xyz/001.webp">
```
They are pointing to non-existent image link. The correct value would be like this:
```html
<meta property="og:image" content="https://tetralog.haipa.xyz/blog/genshin-impact/misc/re-adventure-rank-30/001.webp">
<meta name="twitter:image" content="https://tetralog.haipa.xyz/blog/genshin-impact/misc/re-adventure-rank-30/001.webp">
```

It seems that Hugo wants to look into `static` directory for the `og:image` and `twitter:image`. Isn't it natural to find image from where `index.md` locates first?

## Non-post pages
Similar thing also happens to non-post pages which don't have any `images` front matter.

For example, if I inspect [my blog main page](https://tetralog.haipa.xyz/), its meta is like this:
```html
<meta property="og:image" content="https://tetralog.haipa.xyz/">
<meta name="twitter:image" content="http://tetralog.haipa.xyz/">
```

I thought that this is because I didn't set `params.images` for 'default image' so I set it like this in `params.yml`
```yml
# Default Open Graph / Twitter Cards image
images:
  - images/card-twitter.png
```

<details>
<summary>But I get this weird error message when building site.</summary>

```plaintext
ERROR render of "page" failed: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog@v0.21.0\layouts\blog\single.html:2:5": execute of template failed: template: blog/single.html:2:5: executing "main" at <partial "hb/modules/blog/single" .>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog@v0.21.0\layouts\partials\hb\modules\blog\single.html:32:12": execute of template failed: template: partials/hb/modules/blog/single.html:32:12: executing "partials/hb/modules/blog/single.html" at <partial "hugopress/functions/render-hooks" (dict "Name" "hb-blog-post-content-end" "Page" .)>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hugomods\hugopress@v0.2.2\layouts\partials\hugopress\functions\render-hooks.html:20:8": execute of template failed: template: partials/hugopress/functions/render-hooks.html:20:8: executing "partials/hugopress/functions/render-hooks.html" at <partial .partial $ctx>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog\modules\related-posts@v0.2.4\layouts\partials\hugopress\modules\hb-blog-related-posts\hooks\hb-blog-post-content-end.html:1:3": execute of template failed: template: partials/hugopress/modules/hb-blog-related-posts/hooks/hb-blog-post-content-end.html:1:3: executing "partials/hugopress/modules/hb-blog-related-posts/hooks/hb-blog-post-content-end.html" at <partial "hb/modules/blog-related-posts/index" .Page>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog\modules\related-posts@v0.2.4\layouts\partials\hb\modules\blog-related-posts\index.html:7:13": execute of template failed: template: partials/hb/modules/blog-related-posts/index.html:7:13: executing "partials/hb/modules/blog-related-posts/index.html" at <partial "hb/modules/blog/post/card" (dict "Page" . "Summary" false "Meta" false "Modularize" false)>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog@v0.21.0\layouts\partials\hb\modules\blog\post\card.html:20:7": execute of template failed: template: partials/hb/modules/blog/post/card.html:20:7: executing "partials/hb/modules/blog/post/card.html" at <partial "hb/modules/blog/post/card-img" $page>: error calling partial: execute of template failed: template: partials/hb/modules/blog/post/card-img.html:20:7: executing "partials/hb/modules/blog/post/card-img.html" at <partial "images/image" (dict "Page" $page "Alt" $page.Title "Filename" $img "ClassName" "hb-blog-post-card-img" "Original" false)>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hugomods\images@v0.8.2\layouts\partials\images\image.html:12:7": execute of template failed: template: partials/images/image.html:12:7: executing "partials/images/image.html" at <isset site.Params.images "modern_format">: error calling isset: isset unable to use key of type string as index
Built in 1682 ms
Error: error building site: render: failed to render pages: render of "page" failed: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog@v0.21.0\layouts\blog\single.html:2:5": execute of template failed: template: blog/single.html:2:5: executing "main" at <partial "hb/modules/blog/single" .>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog@v0.21.0\layouts\partials\hb\modules\blog\single.html:32:12": execute of template failed: template: partials/hb/modules/blog/single.html:32:12: executing "partials/hb/modules/blog/single.html" at <partial "hugopress/functions/render-hooks" (dict "Name" "hb-blog-post-content-end" "Page" .)>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hugomods\hugopress@v0.2.2\layouts\partials\hugopress\functions\render-hooks.html:20:8": execute of template failed: template: partials/hugopress/functions/render-hooks.html:20:8: executing "partials/hugopress/functions/render-hooks.html" at <partial .partial $ctx>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog\modules\related-posts@v0.2.4\layouts\partials\hugopress\modules\hb-blog-related-posts\hooks\hb-blog-post-content-end.html:1:3": execute of template failed: template: partials/hugopress/modules/hb-blog-related-posts/hooks/hb-blog-post-content-end.html:1:3: executing "partials/hugopress/modules/hb-blog-related-posts/hooks/hb-blog-post-content-end.html" at <partial "hb/modules/blog-related-posts/index" .Page>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog\modules\related-posts@v0.2.4\layouts\partials\hb\modules\blog-related-posts\index.html:7:13": execute of template failed: template: partials/hb/modules/blog-related-posts/index.html:7:13: executing "partials/hb/modules/blog-related-posts/index.html" at <partial "hb/modules/blog/post/card" (dict "Page" . "Summary" false "Meta" false "Modularize" false)>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hbstack\blog@v0.21.0\layouts\partials\hb\modules\blog\post\card.html:20:7": execute of template failed: template: partials/hb/modules/blog/post/card.html:20:7: executing "partials/hb/modules/blog/post/card.html" at <partial "hb/modules/blog/post/card-img" $page>: error calling partial: execute of template failed: template: partials/hb/modules/blog/post/card-img.html:20:7: executing "partials/hb/modules/blog/post/card-img.html" at <partial "images/image" (dict "Page" $page "Alt" $page.Title "Filename" $img "ClassName" "hb-blog-post-card-img" "Original" false)>: error calling partial: "<LOCALAPPDATA>\hugo_cache\modules\filecache\modules\pkg\mod\github.com\hugomods\images@v0.8.2\layouts\partials\images\image.html:12:7": execute of template failed: template: partials/images/image.html:12:7: executing "partials/images/image.html" at <isset site.Params.images "modern_format">: error calling isset: isset unable to use key of type string as index
```

</details>

What should I do to solve this issue? All I wanted to do was setting default OG/TC image.

---

## TL;DR

* I want to set `og:image` and `twitter:image` to image which is defined in front matter.
* When I try to set `params.images`, Hugo refuses to build site.

**Comment by razonyang** [comment]
This theme uses the built-in template, it's expected, the `images` parameter for `/static` images and external images, if you want to use a page resources (in this case), you'll need to rename the image in pattern `feature*`, `cover*` or `thumbnail*`.

https://github.com/gohugoio/hugo/blob/ac7cffa7e2932fc3c6bd425f86b981dfdef94968/tpl/tplimpl/embedded/templates/opengraph.html#L6-L17

Let's summary up the images selection order:

1. Page's `images` paramerer, which treated as `/static` or external images.
2. Page's resrouces those naming in `feature*`, `cover*` or `thumbnail*` pattern, e.g. `feature.png`, `cover.jpg`.
3. Site's `images` parameter.

**Reply by TetraTheta** [reply]
> 1. Page's `images` parameter, which treated as `/static` or external images.
> 2. Page's resrouces those naming in `feature*`, `cover*` or `thumbnail*` pattern.
> 3. Site's `images` parameter.

Can this behavior be changed like this?

1. Page's `images` parameter (in order)
   1. Absolute image link (includes external image link)
   2. Page resource which its name starts with any of these: `feature`, `cover`, `thumbnail`
   3. [NEW] `images` parameter itself
2. Site's `images` parameter

With this, I can retain how I've defined 'preview image' so far and set OG/TC preview image too.

There are few reason I request this change:

1. I don't always use `001.webp` as preview image.  
   Sometimes, I use `(any string).webp` as preview image.
2. I already have 'structured' shortcode usage.  
   For example, I use this kind of pattern a lot: `{{< gallery/image src="033:034:035" >}}`  
   This will put `033.webp`, `034.webp`, `035.webp` in a row.  
   If any one of these should be changed to something like `cover035.webp`, it would cost a lot of time to find out and replace filename/text for preview image.

In this case, I don't have any other option but copy images used as 'preview image' and rename copied files to `cover.webp`, one by one. And I would have to do same thing whenever I decide to change preview image.

---

So, if I can set OG/TC preview image without copying/changing file name or modifying content of post, it would help me a lot for saving time.

TL;DR

Is it possible to just set OG/TC preview image like this?
```yml
---
images:
  - 001.webp
---
```

**Reply by razonyang** [reply]
> Can this behavior be changed like this?

I'd say no, unless we maintain our Open Graph and Twitter Cards templates or create PR to the built-in template.

Btw, `305cover.png` should be valid as well, the naming pattern position doesn't matter.

**Reply by TetraTheta** [reply]
Either way, it would require me to do a lot of work for both copying (optionally) and editing file names and editing content of post files...

Before:
```yaml
---
images:
  - 030.webp
---
...
{{< gallery/image src="029:030:031" >}}
...
```
```plaintext
<post>
├ index.md
├ 001.webp
├ 002.webp
...
├ 049.webp
└ 050.webp
```
After (Option 1):
```yaml
---
images:
  - 030cover.webp
---
...
{{< gallery/image src="029:030cover:031" >}}
...
```
```plaintext
<post>
├ index.md
├ 001.webp
├ 002.webp
...
├ 030cover.webp
...
├ 049.webp
└ 050.webp
```
After (Option 2):
```yaml
---
images:
  - cover.webp
---
...
{{< gallery/image src="029:030:031" >}}
...
```
```plaintext
<post>
├ index.md
├ cover.webp
├ 001.webp
├ 002.webp
...
├ 049.webp
└ 050.webp
```

This is why I asked for allowing `001.webp` as OG/TC preview image. I have to edit multiple parts of posts, and file names too... :(

I wish if there is a way to automate this...

**Reply by razonyang** [reply]
Will try to create PR to Hugo project those two days.

**Reply by TetraTheta** [reply]
Oh, many thanks!

If I decide to modify and override [`opengraph.html`](https://github.com/gohugoio/hugo/blob/master/tpl/tplimpl/embedded/templates/opengraph.html) and [`twitter_cards.html`](https://github.com/gohugoio/hugo/blob/master/tpl/tplimpl/embedded/templates/twitter_cards.html), should they placed in `<repository>/layouts/partials/_internal/`?

**Reply by razonyang** [reply, marked as answer]
No sure, but you can override those two templates: [layouts/partials/seo/modules/open-graph/index.html](https://github.com/hugomods/seo/blob/main/modules/open-graph/layouts/partials/seo/modules/open-graph/index.html) and [layouts/partials/seo/modules/twitter-cards/index.html](https://github.com/hugomods/seo/blob/main/modules/twitter-cards/layouts/partials/seo/modules/twitter-cards/index.html).

**Reply by TetraTheta** [reply]
Thank you! I should start from there. 

If you create PR to Hugo, can you let me know about it? I want to subscribe that PR.

**Reply by razonyang** [reply]
No problem.

**Reply by TetraTheta** [reply]
As for Open Graph and Twitter Cards, I've managed to modify code to correctly set image URL.

But since I only have two case, so I'm not sure if this can cover all case or not.
- `001.webp`
- `https://blahblah.com/some_image.png`

```html
<!-- Open Graph -->
{{- with $.Params.images -}}
  {{- range first 6 . }}
    {{- $url := "" -}}
    {{- if or (hasPrefix . "http://") (hasPrefix . "https://") -}}
      {{- $url = . -}}
    {{- else -}}
      {{- $url = printf "%s%s" $.Permalink . -}}
    {{ end -}}
    <meta property="og:image" content="{{ $url | absURL }}" />
  {{ end -}}
{{- else -}}
```
```html
<!-- Twitter Cards -->
{{- with $.Params.images -}}
{{- $img := index . 0 -}}
{{- $url := "" -}}
<meta name="twitter:card" content="summary_large_image"/>
{{- if or (hasPrefix $img "http://") (hasPrefix $img "https://") -}}
  {{- $url = $img -}}
{{- else -}}
  {{- $url = printf "%s%s" $.Permalink $img -}}
{{ end -}}
<meta name="twitter:image" content="{{ $url | absURL }}"/>
{{ else -}}
```

**Reply by razonyang** [reply]
> (Option 1) ...

Just noticed this, You shouldn't specify the `images` and the page image resources at the same time in your _ (Option 1)_.
To use page image resources, you've to _remove_ the `images` parameter from front matter, and then rename it in the pattern above, since the `images` gets higher priority.

**Reply by razonyang** [reply]
See https://github.com/gohugoio/hugo/pull/11688, you can find the modified code there.

**Reply by TetraTheta** [reply]
Thank you for letting me know about that! Your code looks much nicer than mine!

**Comment by razonyang** [comment]
The second error was caused by the https://github.com/hugomods/images/issues/26, probably need to rename it to others.

**Reply by razonyang** [reply]
@TetraTheta the second error should be fixed by https://github.com/hugomods/images/releases/tag/v0.8.3.
